### PR TITLE
Fix potential premature finish in Delta CDF function

### PR DESCRIPTION
**Pull request by homar**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java [hunk 1]
```diff
@@ -148,7 +148,11 @@ private TableFunctionProcessorState processCdfFile()
                     currentVersionCommitTimestampAsBlock, page.getPositionCount());
             return TableFunctionProcessorState.Processed.produced(new Page(page.getPositionCount(), resultBlock));
         }
-        return FINISHED;
+        if (deltaLakePageSource.isFinished()) {
+            return FINISHED;
+        } else {
+            return TableFunctionProcessorState.Processed.produced(null);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java:154]: The produced method has a requireNonNull check: https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/function/table/TableFunctionProcessorState.java#L89

**Comment by electrum** [comment]
Commit message title is too long: https://cbea.ms/git-commit/

**Comment by electrum** [comment]
The commit message should be something like "Fix XXX for Delta Lake".

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java [hunk 2]
```diff
@@ -168,7 +172,11 @@ private TableFunctionProcessorState processDataFile()
                     currentVersionCommitTimestampAsBlock, page.getPositionCount());
             return TableFunctionProcessorState.Processed.produced(new Page(page.getPositionCount(), blocks));
         }
-        return FINISHED;
+        if (deltaLakePageSource.isFinished()) {
+            return FINISHED;
+        } else {
```
[the remark this answers is not in the document]
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java]: i will never get used to this format :|

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java (hunk 1 above)
Comment (homar) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java]: yeah sorry for this

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java [hunk 3]
```diff
@@ -135,8 +135,11 @@ public TableFunctionProcessorState process()
 
     private TableFunctionProcessorState processCdfFile()
     {
-        Page page = deltaLakePageSource.getNextPage();
-        if (page != null) {
+        while (!deltaLakePageSource.isFinished()) {
+            Page page = deltaLakePageSource.getNextPage();
+            if (page == null) {
+                continue;
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java:141]: The idea behind `CPS.getNextPage` returning nullable is to break procesisng producing a page takes too long.

I think we should yield here. There are 3 ways we can go about  it

- returning `TableFunctionProcessorState.Blocked.blocked(NOT_BLOCKED)`
- returning `TableFunctionProcessorState.Processed.produced(EMPTY_PAGE)`
- introducing new `TableFunctionProcessorState.Yielded`

let's go with `TableFunctionProcessorState.Processed.produced(EMPTY_PAGE)` for a quick fix
and let's follow up with a PR introducing `TableFunctionProcessorState.Yielded` state

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java [hunk 4]
```diff
@@ -148,7 +150,12 @@ private TableFunctionProcessorState processCdfFile()
                     currentVersionCommitTimestampAsBlock, page.getPositionCount());
             return TableFunctionProcessorState.Processed.produced(new Page(page.getPositionCount(), resultBlock));
         }
-        return FINISHED;
+        if (deltaLakePageSource.isFinished()) {
+            return FINISHED;
+        }
+        else {
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java:156]: redundant `else`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java [hunk 5]
```diff
@@ -168,7 +175,12 @@ private TableFunctionProcessorState processDataFile()
                     currentVersionCommitTimestampAsBlock, page.getPositionCount());
             return TableFunctionProcessorState.Processed.produced(new Page(page.getPositionCount(), blocks));
         }
-        return FINISHED;
+        if (deltaLakePageSource.isFinished()) {
+            return FINISHED;
+        }
+        else {
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java:181]: redundant `else`

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
i tried to understand whether `deltaLakePageSource.getNextPage()` can actually return a null.
I.e. whether we're dealing with a user-facing correctness bug.

this led me to this question: https://github.com/trinodb/trino/pull/13246/files#r1239083632
cc @raunaqmorarka

**Pull request merged**